limit intermediate batch size in nested_loop_join

UBarney opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

part of #16364 
## Rationale for this change
see issue
<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

## What changes are included in this PR?
1. Limit intermediate_batch Size During Filtering
2. Yield Partial Batches on Demand
<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## Are these changes tested?
Yes
<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
3. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

UBarney commented:
# benchmark
I use this [script](https://gist.github.com/UBarney/9dcbf304e65f061d3352b34abd0f0e05#file-sql_bench-py) to do benchmark

| ID | SQL | join_base Time(s) | join_limit_join_batch_size Time(s) | Performance Change |
|----|-----|-------------|------------|-------------------|
| 1 | select t1.value from range(8192) t1 join range(8192) t2 on t1.value + t2.value < t1.value * t2.value; | 0.852 | 0.548 | +35.70% |
| 2 | select t1.value from range(8192) t1 join range(8192) t2 on t1.value + t2.value > t1.value * t2.value; | 0.692 | 0.387 | +44.03% |
| 3 | select t1.value from range(8192) t1 right join range(8192) t2 on t1.value + t2.value > t1.value * t2.value; | 0.707 | 0.386 | +45.35% |
| 4 | select t1.value from range(8192) t1 join range(81920) t2 on t1.value + t2.value < t1.value * t2.value; | Failed | 1.680 | N/A |
| 5 | select t1.value from range(100) t1 join range(819200) t2 on t1.value + t2.value > t1.value * t2.value; | 0.321 | 0.078 | +75.75% |

I'll find out why there is a performance improvement

# memory usage

This version now supports joining large left and right tables, preventing previous OOM errors seen in the main branch

<details>
```
/usr/bin/time -v ./target/release/join_limit_join_batch_size --maxrows 1 
DataFusion CLI v48.0.0
> select t1.value from range(81920) t1 join range(8192) t2 on t1.value + t2.value > t1.value * t2.value;
+-------+
| value |
+-------+
| 1     |
| .     |
| .     |
| .     |
+-------+
180219 row(s) fetched. (First 1 displayed. Use --maxrows to adjust)
Elapsed 4.653 seconds.

        Command being timed: "./target/release/join_limit_join_batch_size --maxrows 1"
        User time (seconds): 2.98
        System time (seconds): 1.67
        Percent of CPU this job got: 24%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:19.00
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 8100012
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 1967478
        Voluntary context switches: 177
        Involuntary context switches: 10
        Swaps: 0
        File system inputs: 0
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0

/usr/bin/time -v ./target/release/join_base --maxrows 1 
DataFusion CLI v48.0.0
> select t1.value from range(81920) t1 join range(8192) t2 on t1.value + t2.value > t1.value * t2.value;
Command terminated by signal 9
        Command being timed: "./target/release/join_base --maxrows 1"
        User time (seconds): 4.04
        System time (seconds): 5.99
        Percent of CPU this job got: 78%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:12.80
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 28959396
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 954
        Minor (reclaiming a frame) page faults: 7216264
        Voluntary context switches: 1181
        Involuntary context switches: 29
        Swaps: 0
        File system inputs: 238720
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```
</details>

jonathanc-n commented:
Those benchmark helper functions are really cool, I'll see if I can take a look today.

UBarney commented:
> I'll find out why there is a performance improvement

From the flame graph (when executing the SQL `select t1.value from range(8192) t1 join range(8192) t2 on t1.value + t2.value < t1.value * t2.value;`) by adjusting the input indices sizes, the execution time of these two functions was reduced

- `apply_join_filter_to_indices` Showed a reduction in execution time (sample count reduced from 528million to 241million).
- `build_batch_from_indices` (excluding the contribution of `apply_join_filter_to_indices`) Showed a reduction in execution time (sample count reduced from 79million to 35million).

But I still can't explain why these two functions performed better. 😂

jonathanc-n commented:
When you are running the benchmarks do they stay consistent?

UBarney commented:
> When you are running the benchmarks do they stay consistent?

Yes. bechmarks result almost consistent.

I ran the benchmarks a few minutes ago on commit.

It's worth noting that the join type in query 3 (q3) was modified from a right join to a left join.

| ID | SQL | join_base Time(s) | join_limit_join_batch_size Time(s) | Performance Change |
|----|-----|-------------|------------|-------------------|
| 1 | select t1.value from range(8192) t1 join range(8192) t2 on t1.value + t2.value < t1.value * t2.value; | 0.821 | 0.559 | +31.89% |
| 2 | select t1.value from range(8192) t1 join range(8192) t2 on t1.value + t2.value > t1.value * t2.value; | 0.633 | 0.367 | +41.99% |
| 3 | select t1.value from range(8192) t1 left join range(8192) t2 on t1.value + t2.value > t1.value * t2.value; | 0.632 | 0.365 | +42.29% |
| 4 | select t1.value from range(8192) t1 join range(81920) t2 on t1.value + t2.value < t1.value * t2.value; | Failed | 1.495 | N/A |
| 5 | select t1.value from range(100) t1 join range(819200) t2 on t1.value + t2.value > t1.value * t2.value; | 0.285 | 0.072 | +74.81% |
| 6 | select t1.value from range(100) t1 join range(819200) t2 on t1.value + t2.value < t1.value * t2.value; | 0.392 | 0.172 | +56.13% |


| SQL Query | join_base Memory | join_limit_join_batch_size Memory | Improvement |
|-----------|----------------|---------------------------------|-------------|
| select t1.value from range(8192) t1 join range(8192) t2 on t1.value + t2.value < t1.value * t2.value; | 2.82 GB | 1.57 GB | ↓ 44.2% |
| select t1.value from range(8192) t1 join range(8192) t2 on t1.value + t2.value > t1.value * t2.value; | 2.83 GB | 857.4 MB | ↓ 70.4% |
| select t1.value from range(8192) t1 left join range(8192) t2 on t1.value + t2.value > t1.value * t2.value; | 2.82 GB | 847.6 MB | ↓ 70.7% |
| select t1.value from range(8192) t1 join range(81920) t2 on t1.value + t2.value < t1.value * t2.value; | 26.51 GB ❌ (exit code: 137) | 15.11 GB | N/A (process failed) |
| select t1.value from range(100) t1 join range(819200) t2 on t1.value + t2.value > t1.value * t2.value; | 1008.6 MB | 330.4 MB | ↓ 67.2% |
| select t1.value from range(100) t1 join range(819200) t2 on t1.value + t2.value < t1.value * t2.value; | 1.35 GB | 652.1 MB | ↓ 52.7% |

UBarney commented:
> * `apply_join_filter_to_indices` Showed a reduction in execution time (sample count reduced from 528million to 241million).

The benchmark results indicate that restricting the row count of `intermediate_batch` in `apply_join_filter_to_indices` indeed enhances performance.
```
Benchmarking nlj/filter/batch_size==8192: Collecting 100 samples in estimated 156.94 s (30
nlj/filter/batch_size==8192
                        time:   [487.31 ms 488.41 ms 489.68 ms]
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  4 (4.00%) high severe
Benchmarking nlj/filter/batch_size==8192*8192: Collecting 100 samples in estimated 126.81 
nlj/filter/batch_size==8192*8192
                        time:   [658.42 ms 659.90 ms 661.49 ms]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild
```

https://gist.github.com/UBarney/23fdb597f43bfcffe4f781fb6b99e579

jonathanc-n commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -828,13 +833,127 @@ impl<T: BatchTransformer> NestedLoopJoinStream<T> {
+    fn get_next_join_result(&mut self) -> Result<Option<RecordBatch>> {
+        let (left_indices, right_indices, start) =
+            self.join_result_status.as_mut().ok_or_else(|| {
+                datafusion_common::_internal_datafusion_err!(
+                    "should have join_result_status"
```
I think we can change this to be more verbose, and we can use 
`internal_err!`
```suggestion
                internal_err!(
                    "get_next_join_result called without initializing join_result_status"
```

UBarney replied:
> we can use internal_err!


`internal_err!` return `Result` but `ok_or_else` require closure return `error`

https://github.com/rust-lang/rust/blob/86e05cd300fac9e83e812c4d46582b48db780d8f/library/core/src/option.rs#L1334-L1336

jonathanc-n reviewed (commented):
I'm a bit confused where the performance increase is coming from as well. I noted down some nits; I'll take a better look tonight

korowa commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -883,44 +1002,66 @@ impl<T: BatchTransformer> NestedLoopJoinStream<T> {
+        if self.join_result_status.is_none() {
+            self.join_metrics.input_batches.add(1);
+            self.join_metrics.input_rows.add(batch.num_rows());
+            let _timer = self.join_metrics.join_time.timer();
```
Maybe only one timer covering all the function will be enough (instead on this timer, and the one on L1023)?

UBarney replied:
> Maybe only one timer covering all the function will be enough (instead on this timer, and the one on L1023)?

That was my initial approach, but it caused a borrow checker error (E0502).

The issue is a conflict between an immutable borrow for the timer (`self.join_metrics.join_time`) and a mutable borrow required by `self.get_next_join_result`.

We can create the timer inside `get_next_join_result` itself, rather than in the caller

2010YOUY01 replied:
I think you can clone it, the underlying structure of this metric is `Arc<AtomicUsize>`, so the cloned version points to the same counter.

UBarney replied:
Done. I've updated the code to clone `join_metrics.join_time`

UBarney commented on datafusion/physical-plan/src/joins/utils.rs:
```diff
@@ -843,24 +844,56 @@ pub(crate) fn apply_join_filter_to_indices(
-        .expression()
-        .evaluate(&intermediate_batch)?
-        .into_array(intermediate_batch.num_rows())?;
+    let filter_result = if let Some(max_size) = max_intermediate_size {
```
>Can we enforce it before filtering, while calculating build/probe_indices args for this function (in NestedLoopJoinExec::build_join_indices)?

I'll do it in next pr. https://github.com/apache/datafusion/issues/16364#issuecomment-2975520489

> Why batch_size enforcement should take place during filtering

1. Although the "Process the Cartesian Product Incrementally" step is designed to limit the input size for `apply_join_filter_to_indices`, the size of a single batch can still be very large (up to `left_table.now_rows() * N`). When the left table itself is large, this can lead to the creation of a large `record_batch`.
2. Benchmarks indicate that executing joins is faster with this enforcement in place. https://github.com/apache/datafusion/pull/16443#issuecomment-2993893069

UBarney replied:
> Shouldn't we have this done in this pull request?

Yes, Sorry, I misunderstood.

After moving this logic to `build_join_indices`, the performance got worse. Is it possible that there's an extra `concat` operation?


| ID | SQL | join_limit_join_batch_size Time(s) | mv_limit_to_build Time(s) | Performance Change |
|----|-----|-------------|------------|-------------------|
| 1 | select t1.value from range(8192) t1 join range(8192) t2 on t1.value + t2.value < t1.value * t2.value; | 0.560 | 0.639 | -14.03% |
| 2 | select t1.value from range(8192) t1 join range(8192) t2 on t1.value + t2.value > t1.value * t2.value; | 0.389 | 0.370 | +4.78% |
| 3 | select t1.value from range(8192) t1 left join range(8192) t2 on t1.value + t2.value > t1.value * t2.value; | 0.373 | 0.372 | +0.11% |
| 4 | select t1.value from range(8192) t1 join range(81920) t2 on t1.value + t2.value < t1.value * t2.value; | 1.552 | 2.162 | -39.30% |
| 5 | select t1.value from range(100) t1 join range(819200) t2 on t1.value + t2.value > t1.value * t2.value; | 0.061 | 0.061 | +0.98% |
| 6 | select t1.value from range(100) t1 join range(819200) t2 on t1.value + t2.value < t1.value * t2.value; | 0.156 | 0.197 | -26.41% |

<details>

```
   if let Some(filter) = filter {
        let num_indices = left_indices.len();
        let mut left_chunks =
            Vec::with_capacity(num_indices.div_ceil(max_intermediate_batch_size));
        let mut right_chunks =
            Vec::with_capacity(num_indices.div_ceil(max_intermediate_batch_size));

        for i in (0..num_indices).step_by(max_intermediate_batch_size) {
            let end = min(num_indices, i + max_intermediate_batch_size);
            let len = end - i;

            let left_slice = left_indices.slice(i, len);
            let right_slice = right_indices.slice(i, len);

            let (left_filtered, right_filtered) = apply_join_filter_to_indices(
                left_batch,
                right_batch,
                left_slice,
                right_slice,
                filter,
                JoinSide::Left,
                None,
            )?;

            left_chunks.push(left_filtered);
            right_chunks.push(right_filtered);
        }

        let left_ref: Vec<&dyn Array> =
            left_chunks.iter().map(|x| x as &dyn Array).collect();

        let right_ref: Vec<&dyn Array> =
            right_chunks.iter().map(|x| x as &dyn Array).collect();

        Ok((
            downcast_array(compute::concat(&left_ref)?.as_ref()),
            downcast_array(compute::concat(&right_ref)?.as_ref()),
        ))

    }
```

</details>

I think it's better to apply this limit within `apply_join_filter_to_indices`. The reason is that this function might be called by others like hash_join, and the `(left_indices, right_indices)` they pass in could also be very large.

UBarney commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -828,13 +833,127 @@ impl<T: BatchTransformer> NestedLoopJoinStream<T> {
+
+        let current_start = *start;
+
+        if left_indices.is_empty() && right_indices.is_empty() && current_start == 0 {
```
That was my initial approach. However, it resulted in an output with 0 rows and 0 columns, which seems to be incorrect and caused the test to fail.

You can see the failed CI run here:
https://github.com/apache/datafusion/actions/runs/15734253347/job/44343070926?pr=16443#step:5:1208

jonathanc-n replied:
Can we just build and return an empty batch instead of calling build_batch_from_indices?

2010YOUY01 replied:
I don't get this `status.processed_count = 1` logic either, perhaps you can add a quick comment to explain it?

UBarney replied:
Now I understand why [this test](https://github.com/apache/datafusion/blob/acf0bbe1cedf3fe0155de5f41a5b66046e262e04/datafusion/core/src/dataframe/mod.rs#L1250-L1263) passed after I changed the return value from `None` to `RecordBatch::new_empty`.

In this unit test, the join result is converted to a string and then compared with the expected output. When converting to a string, it retrieves the schema from the `record_batch` (as the passed `schema_opt` is `None`).
https://github.com/apache/arrow-rs/blob/7b219f98c25fcd318a0c207f51a41398d1b23724/arrow-cast/src/pretty.rs#L186

When executed in the CLI, there's no issue even if the Nested Loop Join (NLJ) returns 0 record batches.

```
> select t1.value from range(1) t1 join range(1) t2 on t1.value + t2.value >100;

+-------+
| value |
+-------+
+-------+
0 row(s) fetched. 
```

From a compatibility standpoint, I think it's better to keep it consistent with the previous behavior.

UBarney commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -1215,16 +1324,25 @@ pub(crate) mod tests {
+    fn new_task_ctx() -> Arc<TaskContext> {
+        let base = TaskContext::default();
+        // limit max size of intermediate batch used in nlj to 1
+        let cfg = base.session_config().clone().with_batch_size(1);
```
DONE

jonathanc-n reviewed (commented):
Thanks @UBarney, just some comments

jonathanc-n commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -729,10 +716,26 @@ struct NestedLoopJoinStream<T> {
+    // - probe_indices: row indices from probe-side table (right table)
+    // - processed_count: number of index pairs already processed into output batches
+    // We have completed join result for indices [0..processed_count)
+    join_result_status: Option<(
```
I still think we should make this into a struct

jonathanc-n commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -828,13 +828,127 @@ impl<T: BatchTransformer> NestedLoopJoinStream<T> {
+            .batch();
+
+        let right_batch = match &self.state {
+            NestedLoopJoinStreamState::ProcessProbeBatch(record_batch) => record_batch,
+            NestedLoopJoinStreamState::OutputUnmatchedBuildRows(record_batch) => {
```
```suggestion
            NestedLoopJoinStreamState::ProcessProbeBatch(record_batch) | NestedLoopJoinStreamState::OutputUnmatchedBuildRows(record_batch) => {
```

jonathanc-n commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -828,13 +828,127 @@ impl<T: BatchTransformer> NestedLoopJoinStream<T> {
+            }
+            _ => {
+                return internal_err!(
+                    "state should be ProcessProbeBatch or OutputUnmatchBatch"
```
```suggestion
                    "State should be ProcessProbeBatch or OutputUnmatchedBuildRows"
```

jonathanc-n commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -883,44 +1000,63 @@ impl<T: BatchTransformer> NestedLoopJoinStream<T> {
+                }
             }
+        } else {
+            internal_err!("state should be OutputUnmatchBatch")
```
```suggestion
            internal_err!("State should be OutputUnmatchedBuildRows")
```

UBarney commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -883,44 +1000,63 @@ impl<T: BatchTransformer> NestedLoopJoinStream<T> {
+    fn build_unmatched_output(
+        &mut self,
+    ) -> Result<StatefulStreamResult<Option<RecordBatch>>> {
+        if matches!(
```
Yes, we don't need this check

jonathanc-n reviewed (approved):
Thanks @UBarney This looks good to me! Looking forward to reviewing the follow up prs

jonathanc-n commented:
@korowa @2010YOUY01 Are you able to take a quick look? Thanks!

Dandandan commented on datafusion/physical-plan/src/joins/utils.rs:
```diff
@@ -843,24 +844,56 @@ pub(crate) fn apply_join_filter_to_indices(
+        let filter_refs: Vec<&dyn Array> =
+            filter_results.iter().map(|a| a.as_ref()).collect();
+
+        compute::concat(&filter_refs)?
```
Would be nice to avoid this and rely on `CoalesceBatches` instead

Dandandan replied:
oh wait this is filter, I have another suggestion

alamb commented on datafusion/physical-plan/src/joins/utils.rs:
```diff
@@ -843,24 +844,56 @@ pub(crate) fn apply_join_filter_to_indices(
+                .expression()
+                .evaluate(&intermediate_batch)?
+                .into_array(intermediate_batch.num_rows())?;
+            filter_results.push(filter_result);
```
coalesce is now available in datafusion (we have upgraded to a new arrow version)

I hope to continue improving coalesce over time (especially for this common usecase of building up the output of filter)

alamb replied:
Speciically -- https://docs.rs/arrow/latest/arrow/compute/struct.BatchCoalescer.html

UBarney replied:
However, after replacing `concat` with `coalescer.push_batch_with_filter`, performance actually decreased.

<details>
<summary>code</summary>

```
pub(crate) fn apply_join_filter_to_indices(
    build_input_buffer: &RecordBatch,
    probe_batch: &RecordBatch,
    build_indices: UInt64Array,
    probe_indices: UInt32Array,
    filter: &JoinFilter,
    build_side: JoinSide,
    max_intermediate_size: Option<usize>,
) -> Result<(UInt64Array, UInt32Array)> {
    if build_indices.is_empty() && probe_indices.is_empty() {
        return Ok((build_indices, probe_indices));
    };

    if let Some(max_size) = max_intermediate_size {
        let indices_schema = Arc::new(Schema::new(vec![
            Field::new("build_indices", arrow::datatypes::DataType::UInt64, false),
            Field::new("probe_indices", arrow::datatypes::DataType::UInt32, false),
        ]));
        let build_indices = Arc::new(build_indices);
        let probe_indices = Arc::new(probe_indices);

        let indices_batch = RecordBatch::try_new(
            indices_schema,
            vec![
                Arc::clone(&build_indices) as Arc<dyn Array>,
                Arc::clone(&probe_indices) as Arc<dyn Array>,
            ],
        )?;

        let mut coalescer =
            BatchCoalescer::new(indices_batch.schema(), indices_batch.num_rows());

        for i in (0..build_indices.len()).step_by(max_size) {
            let end = min(build_indices.len(), i + max_size);
            let len = end - i;
            let intermediate_batch = build_batch_from_indices(
                filter.schema(),
                build_input_buffer,
                probe_batch,
                &build_indices.slice(i, len),
                &probe_indices.slice(i, len),
                filter.column_indices(),
                build_side,
            )?;
            let filter_result = filter
                .expression()
                .evaluate(&intermediate_batch)?
                .into_array(intermediate_batch.num_rows())?;

            coalescer.push_batch_with_filter(
                indices_batch.slice(i, len),
                as_boolean_array(&filter_result)?,
            )?;
        }

        coalescer.finish_buffered_batch()?;

        let result = coalescer.next_completed_batch();
        if result.is_none() {
            return Ok((build_indices.slice(0, 0), probe_indices.slice(0, 0)));
        }

        if coalescer.has_completed_batch() {
            return internal_err!("should not have completed_batch");
        }

        let (_, arrays, _) = result.unwrap().into_parts();
        return Ok((
            downcast_array(arrays[0].as_ref()),
            downcast_array(arrays[1].as_ref()),
        ));
    }
    
    let intermediate_batch = build_batch_from_indices(
        filter.schema(),
        build_input_buffer,
        probe_batch,
        &build_indices,
        &probe_indices,
        filter.column_indices(),
        build_side,
    )?;

    let filter_result = filter
        .expression()
        .evaluate(&intermediate_batch)?
        .into_array(intermediate_batch.num_rows())?;

    let mask = as_boolean_array(&filter_result)?;

    let left_filtered = compute::filter(&build_indices, mask)?;
    let right_filtered = compute::filter(&probe_indices, mask)?;
    Ok((
        downcast_array(left_filtered.as_ref()),
        downcast_array(right_filtered.as_ref()),
    ))
}
```

</details>


<details>

<summary>bench result</summary>


| ID | SQL | join_limit_join_batch_size Time(s) | use_BatchCoalescer Time(s) | Performance Change |
|----|-----|-------------|------------|-------------------|
| 1 | select t1.value from range(8192) t1 join range(8192) t2 on t1.value + t2.value < t1.value * t2.value; | 0.559 | 0.671 | 1.20x slower 🐌 |
| 2 | select t1.value from range(8192) t1 join range(8192) t2 on t1.value + t2.value > t1.value * t2.value; | 0.377 | 0.371 | +1.02x faster 🚀 |
| 3 | select t1.value from range(8192) t1 left join range(8192) t2 on t1.value + t2.value > t1.value * t2.value; | 0.363 | 0.363 | +1.00x faster 🚀 |
| 4 | select t1.value from range(8192) t1 join range(81920) t2 on t1.value + t2.value < t1.value * t2.value; | 1.556 | 2.031 | 1.30x slower 🐌 |
| 5 | select t1.value from range(100) t1 join range(819200) t2 on t1.value + t2.value > t1.value * t2.value; | 0.063 | 0.057 | +1.11x faster 🚀 |
| 6 | select t1.value from range(100) t1 join range(819200) t2 on t1.value + t2.value < t1.value * t2.value; | 0.153 | 0.194 | 1.27x slower 🐌 |

| SQL Query                                                                                                  | join_limit_join_batch_size Memory | use_BatchCoalescer Memory | Improvement     |
| ---------------------------------------------------------------------------------------------------------- | --------------------------------- | ------------------------- | --------------- |
| select t1.value from range(8192) t1 join range(8192) t2 on t1.value + t2.value < t1.value * t2.value;      | 1.57 GB                           | 2.31 GB                   | 1.47x more 🐌   |
| select t1.value from range(8192) t1 join range(8192) t2 on t1.value + t2.value > t1.value * t2.value;      | 841.5 MB                          | 824.9 MB                  | +1.02x saved 🚀 |
| select t1.value from range(8192) t1 left join range(8192) t2 on t1.value + t2.value > t1.value * t2.value; | 845.3 MB                          | 824.6 MB                  | +1.03x saved 🚀 |
| select t1.value from range(8192) t1 join range(81920) t2 on t1.value + t2.value < t1.value * t2.value;     | 15.00 GB                          | 20.36 GB                  | 1.36x more 🐌   |
| select t1.value from range(100) t1 join range(819200) t2 on t1.value + t2.value > t1.value * t2.value;     | 328.1 MB                          | 327.6 MB                  | +1.00x saved 🚀 |
| select t1.value from range(100) t1 join range(819200) t2 on t1.value + t2.value < t1.value * t2.value;     | 659.8 MB                          | 810.4 MB                  | 1.23x more 🐌   |

</details>

Dandandan replied:
I think batch coalescer won't make this faster as this is buffering everything in memory anyway. 

The main idea would be to apply the filters iteratively to the incoming `RecordBatch` instead of the indices, so we have to change the API / implementation a bit more.

It's fine to leave this as future change.

2010YOUY01 replied:
Yes, that's expected, now we only got the interface ready, the efficient implementation is still WIP.
See https://docs.rs/arrow-select/55.2.0/src/arrow_select/coalesce.rs.html#194

2010YOUY01 commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -883,44 +996,60 @@ impl<T: BatchTransformer> NestedLoopJoinStream<T> {
+        match join_result {
+            Some(res) => {
+                self.join_metrics.output_batches.add(1);
+                self.join_metrics.output_rows.add(res.num_rows());
```
we don't have to count output_rows here: it would be automatically counted in the outer poll

This was made by a recent change: https://github.com/apache/datafusion/pull/16500
While output_batches still needs to be manually tracked here, it could also be automatically counted in the future.

2010YOUY01 commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -883,44 +996,60 @@ impl<T: BatchTransformer> NestedLoopJoinStream<T> {
+    fn build_unmatched_output(
+        &mut self,
+    ) -> Result<StatefulStreamResult<Option<RecordBatch>>> {
+        let start = Instant::now();
```
nit: I think we can just construct a timer guard here and let it stop on drops.

2010YOUY01 commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -689,6 +674,8 @@ enum NestedLoopJoinStreamState {
     ProcessProbeBatch(RecordBatch),
     /// Indicates that probe-side has been fully processed
     ExhaustedProbeSide,
+    /// Output unmatched build-side rows
```
```suggestion
    /// Output unmatched build-side rows.
    /// The indices for rows to output has already been calculated in the previous
    /// `ExhaustedProbeSide` state. In this state the final batch will be materialized
    // incrementally.
    // The inner `RecordBatch` is an empty dummy batch used to get right schema.
```
Maybe we can also rename `ExhaustedProbeSide` to `PrepareUnmatchedBuildRows` to be more accurate.

2010YOUY01 reviewed (commented):
I think this PR's idea is great, the implementation overall looks good to me.

I recommend to doc more high-level ideas to key functions, to make this module easier to maintain in the future, specifically: `build_unmatched_output()`, `prepare_unmatched_output_indices()`, and `get_next_join_result()`

UBarney commented:
I have addressed all of your comments. @2010YOUY01  please take another look



> I recommend to doc more high-level ideas to key functions, to make this module easier to maintain in the future, specifically: `build_unmatched_output()`, `prepare_unmatched_output_indices()`, and `get_next_join_result()`

Since the logic of `build_unmatched_output` is quite straightforward, I don't think it's necessary to add extra documentation.
I've now added documentation for the other two functions

2010YOUY01 commented:
> I have addressed all of your comments. @2010YOUY01 please take another look
> 
> > I recommend to doc more high-level ideas to key functions, to make this module easier to maintain in the future, specifically: `build_unmatched_output()`, `prepare_unmatched_output_indices()`, and `get_next_join_result()`
> 
> Since the logic of `build_unmatched_output` is quite straightforward, I don't think it's necessary to add extra documentation. I've now added documentation for the other two functions

I'm starting a second pass. I haven't fully grasped the internal logic of `get_next_join_result()` for the special join types yet, but I’ll continue soon.

I think documenting the semantics of left/right indices for different join types can help readability, like
- For inner joins: left and right indices should all be valid
- For left/right/full joins: left and right indices can use null to represent unmatched rows
- Question: do we have other special semantics for semi/anti/mark joins?

jonathanc-n commented:
@2010YOUY01 Special types need to only return the matching rows, so only one side needs to return rows while the other side can return a null array and not be projected in the final result. This functionality was only moved and had already existed before this pull request where most of it sits in `build_batch_from_indices`

2010YOUY01 commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -828,15 +845,125 @@ impl<T: BatchTransformer> NestedLoopJoinStream<T> {
+            return Ok(Some(res));
+        }
+
+        if matches!(self.join_type, JoinType::RightSemi | JoinType::RightAnti) {
```
I think from here to the end of the function, it can look nicer if we structure it like this way
```rust
match self.join_type {
    JoinType::RightSemi | JoinType::RightAnti => {...}
    JoinType::RightMark => {...}
    JoinType::......(others) => {}
    _ => {unreachable!()}
}
```

2010YOUY01 commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -705,8 +696,29 @@ impl NestedLoopJoinStreamState {
+/// - Poll 1: output batch[0..100], processed_count = 100  
+/// - Poll 2: output batch[100..200], processed_count = 200
+/// - ...continues until processed_count = 5000
+struct JoinResultStatus {
```
nit: `Status` is most commonly used for error code/ state flags, perhaps we can `JoinResultProgress` here to avoid confusion?

2010YOUY01 reviewed (approved):
LGTM, thanks again. 🙏🏼

UBarney commented:
> * Refactor so we limit building the entire cartesian product of both batches (this is already covered in the issue and I believe @UBarney is willing to work on this)

Yes. I'll work on this very soon

2010YOUY01 commented:
> select t1.value from range(100) t1 join range(819200) t2 on t1.value + t2.value < t1.value * t2.value;

I'm happy to include this benchmark in the bench suite this week, unless you were already planning to add it yourself @UBarney

UBarney commented on datafusion/physical-plan/src/joins/utils.rs:
```diff
@@ -843,24 +844,56 @@ pub(crate) fn apply_join_filter_to_indices(
+                filter.column_indices(),
+                build_side,
+            )?;
+            let filter_result = filter
```
This optimization boosts performance with a much higher IPC of 1.75 (vs 0.87), achieved by dramatically cutting LLC misses from 109M to 25M, even with a similar L1 miss rate.

<details>

```
	sudo perf stat  -e cycles,instructions,L1-dcache-load-misses,L1-dcache-loads,LLC-loads,LLC-load-misses ./limit_batch_size@36991aca -c 'select t1.value from range(100) t1 join range(819200) t2 on (t1.value + t2.value) % 1000 = 0;	' --maxrows 1
	
	sudo perf stat  -e cycles,instructions,L1-dcache-load-misses,L1-dcache-loads,LLC-loads,LLC-load-misses ./join_base@6965fd32 -c 'select t1.value from range(100) t1 join range(819200) t2 on (t1.value + t2.value) % 1000 = 0;	' --maxrows 1
	
	DataFusion CLI v48.0.0
	+-------+
	| value |
	+-------+
	| 40    |
	| .     |
	| .     |
	| .     |
	+-------+
	81901 row(s) fetched. (First 1 displayed. Use --maxrows to adjust)
	Elapsed 0.067 seconds.
	
	
	 Performance counter stats for './limit_batch_size@36991aca -c select t1.value from range(100) t1 join range(819200) t2 on (t1.value + t2.value) % 1000 = 0;        --maxrows 1':
	
	        1901401922      cycles
	        3325776634      instructions                     #    1.75  insn per cycle
	          32419611      L1-dcache-load-misses            #    5.27% of all L1-dcache accesses
	         614645891      L1-dcache-loads
	   <not supported>      LLC-loads
	   <not supported>      LLC-load-misses
	
	       0.073244586 seconds time elapsed
	
	       0.448238000 seconds user
	       0.044823000 seconds sys
	
	
	DataFusion CLI v48.0.0
	+-------+
	| value |
	+-------+
	| 99    |
	| .     |
	| .     |
	| .     |
	+-------+
	81901 row(s) fetched. (First 1 displayed. Use --maxrows to adjust)
	Elapsed 0.131 seconds.
	
	
	 Performance counter stats for './join_base@6965fd32 -c select t1.value from range(100) t1 join range(819200) t2 on (t1.value + t2.value) % 1000 = 0;       --maxrows 1':
	
	        3696196789      cycles
	        3201132508      instructions                     #    0.87  insn per cycle
	          21781750      L1-dcache-load-misses            #    3.68% of all L1-dcache accesses
	         592094439      L1-dcache-loads
	   <not supported>      LLC-loads
	   <not supported>      LLC-load-misses
	
	       0.139081088 seconds time elapsed
	
	       0.835575000 seconds user
	       0.111277000 seconds sys

	(venv) √ devhomeinsp  ~/c/d/t/release > valgrind --cache-sim=yes  --tool=cachegrind ./join_base@6965fd32 -c 'select t1.value from range(8192) t1 join range(8192) t2 on t1.value + t2.value > t1.value * t2.value;' --maxrows 1
	
	==94454== Cachegrind, a high-precision tracing profiler
	==94454== Copyright (C) 2002-2017, and GNU GPL'd, by Nicholas Nethercote et al.
	==94454== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
	==94454== Command: ./join_base@6965fd32 -c select\ t1.value\ from\ range(8192)\ t1\ join\ range(8192)\ t2\ on\ t1.value\ +\ t2.value\ \>\ t1.value\ *\ t2.value; --maxrows 1
	==94454== 
	--94454-- warning: L3 cache found, using its data for the LL simulation.
	--94454-- warning: specified LL cache: line_size 64  assoc 12  total_size 31,457,280
	--94454-- warning: simulated LL cache: line_size 64  assoc 15  total_size 31,457,280
	DataFusion CLI v48.0.0
	+-------+
	| value |
	+-------+
	| 1     |
	| .     |
	| .     |
	| .     |
	+-------+
	32763 row(s) fetched. (First 1 displayed. Use --maxrows to adjust)
	Elapsed 7.948 seconds.
	
	==94454== 
	==94454== I refs:        3,555,994,712
	==94454== I1  misses:           66,444
	==94454== LLi misses:           26,028
	==94454== I1  miss rate:          0.00%
	==94454== LLi miss rate:          0.00%
	==94454== 
	==94454== D refs:          813,250,121  (475,263,085 rd   + 337,987,036 wr)
	==94454== D1  misses:      118,285,307  ( 71,937,864 rd   +  46,347,443 wr)
	==94454== LLd misses:      109,455,796  ( 63,122,399 rd   +  46,333,397 wr)
	==94454== D1  miss rate:          14.5% (       15.1%     +        13.7%  )
	==94454== LLd miss rate:          13.5% (       13.3%     +        13.7%  )
	==94454== 
	==94454== LL refs:         118,351,751  ( 72,004,308 rd   +  46,347,443 wr)
	==94454== LL misses:       109,481,824  ( 63,148,427 rd   +  46,333,397 wr)
	==94454== LL miss rate:            2.5% (        1.6%     +        13.7%  )
	
	
	
	(venv) √ devhomeinsp  ~/c/d/t/release > valgrind --cache-sim=yes --tool=cachegrind ./limit_batch_size@36991aca -c 'select t1.value from range(8192) t1 join range(8192) t2 on t1.value + t2.value > t1.value * t2.value;' --maxrows 1
	
	==96086== Cachegrind, a high-precision tracing profiler
	==96086== Copyright (C) 2002-2017, and GNU GPL'd, by Nicholas Nethercote et al.
	==96086== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
	==96086== Command: ./limit_batch_size@36991aca -c select\ t1.value\ from\ range(8192)\ t1\ join\ range(8192)\ t2\ on\ t1.value\ +\ t2.value\ \>\ t1.value\ *\ t2.value; --maxrows 1
	==96086== 
	--96086-- warning: L3 cache found, using its data for the LL simulation.
	--96086-- warning: specified LL cache: line_size 64  assoc 12  total_size 31,457,280
	--96086-- warning: simulated LL cache: line_size 64  assoc 15  total_size 31,457,280
	DataFusion CLI v48.0.0
	+-------+
	| value |
	+-------+
	| 1     |
	| .     |
	| .     |
	| .     |
	+-------+
	32763 row(s) fetched. (First 1 displayed. Use --maxrows to adjust)
	Elapsed 8.163 seconds.
	
	==96086== 
	==96086== I refs:        3,663,944,959
	==96086== I1  misses:          944,257
	==96086== LLi misses:           27,378
	==96086== I1  miss rate:          0.03%
	==96086== LLi miss rate:          0.00%
	==96086== 
	==96086== D refs:          847,265,289  (495,073,876 rd   + 352,191,413 wr)
	==96086== D1  misses:      122,392,985  ( 74,620,328 rd   +  47,772,657 wr)
	==96086== LLd misses:       25,750,761  ( 12,815,239 rd   +  12,935,522 wr)
	==96086== D1  miss rate:          14.4% (       15.1%     +        13.6%  )
	==96086== LLd miss rate:           3.0% (        2.6%     +         3.7%  )
	==96086== 
	==96086== LL refs:         123,337,242  ( 75,564,585 rd   +  47,772,657 wr)
	==96086== LL misses:        25,778,139  ( 12,842,617 rd   +  12,935,522 wr)
	==96086== LL miss rate:            0.6% (        0.3%     +         3.7%  )
```

</details>
